### PR TITLE
add methods to v2/lazy_frame that do caching

### PR DESCRIPTION
**Pull request by Raynos**
Most of these changes are straight forward and
should improve perf.

The change to the header parsing hot code might
reduce perf.. not really sure, need more full
benchmarks.

I tried to microbench my change to header parsing
to make sure the new method still get's inlined.

I got:

```
raynos at raynos-ThinkPad-T440p  ~/uber/tchannel-node on lazy-caching*
$ cat inline | grep 'findOffset called'
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Inlined findOffset called from getValue.
Did not inline findOffset called from getValue (target not inlineable).
```

Not sure why it bailed the inline out...

r: @jcorbin @rf

### CODE REVIEW

File: v2/args.js [hunk 1]
```diff
@@ -62,6 +62,10 @@ ArgRW.prototype.readFrom = function readFrom(buffer, offset) {
     return this.bufrw.readFrom(buffer, offset);
 };
 
+ArgRW.prototype.readFromstr = function readFromstr(buffer, offset) {
+    return this.strrw.readFrom(buffer, offset);
```
Comment (jcorbin) [inline v2/args.js]: please just inline

File: v2/call.js [hunk 2]
```diff
@@ -125,6 +154,51 @@ CallRequest.RW.lazy.readArg1 = function readArg1(frame, headers) {
...
+    if (res.err) {
+        return res;
+    }
+    offset = res.offset;
+
+    // READ arg~2
+    return argsrw.argrw.readFrom(frame.buffer, offset);
+};
+
+CallRequest.RW.lazy.readArg1str = function readArg1str(frame, headers) {
```
Comment (jcorbin) [inline v2/call.js]: please `readArg1Str`

File: v2/call.js [hunk 3]
```diff
@@ -125,6 +154,51 @@ CallRequest.RW.lazy.readArg1 = function readArg1(frame, headers) {
...
+    if (frame.cache.arg1Str) {
+        return frame.cache.arg1Str;
+    }
+
+    var res = null;
+    var offset = 0;
+
+    // TODO: memoize computed offsets on frame between readService, readArg1,
```
Comment (jcorbin) [inline v2/call.js]: I bet you that this caching would be quite impactful to hyperbahn

**Comment by Raynos** [comment]
Ok I did **way** more then just caching.

I added a benchmark... and then made it 400% faster.

How did I make it 400% faster ? INLINE EVERYTHING. HAND OPTIMIZE EVERYTHING.

File: benchmarks/micro/lazy-headers-perf.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Copyright (c) 2015 Uber Technologies, Inc.
```
Comment (jcorbin) [inline benchmarks/micro/lazy-headers-perf.js]: double double copy copy right right

File: benchmarks/micro/reading-important-fields.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Copyright (c) 2015 Uber Technologies, Inc.
```
Comment (jcorbin) [inline benchmarks/micro/reading-important-fields.js]: double double copy copy right right

File: v2/call.js [hunk 6]
```diff
@@ -35,11 +37,30 @@ var Frame = require('./frame');
...
 
+var CN_BUFFER = new Buffer('cn');
+var CN_VALUE = CN_BUFFER.readUInt16BE(0, false);
+
 var ResponseCodes = {
     OK: 0x00,
     Error: 0x01
 };
 
+var fastBufferToString = process.version === 'v0.10.32' ?
```
Comment (jcorbin) [inline v2/call.js]: :disappointed:

File: v2/call.js [hunk 7]
```diff
@@ -96,37 +117,226 @@ CallRequest.RW.lazy.readTracing = function lazyReadTracing(frame) {
 
 CallRequest.RW.lazy.serviceOffset = CallRequest.RW.lazy.tracingOffset + 25;
 CallRequest.RW.lazy.readService = function lazyReadService(frame) {
+    if (frame.cache.serviceRes) {
```
Comment (jcorbin) [inline v2/call.js]: better to `!== null`?

File: v2/call.js [hunk 8]
```diff
@@ -96,37 +117,226 @@ CallRequest.RW.lazy.readTracing = function lazyReadTracing(frame) {
...
+        return frame.cache.serviceRes;
+    }
     // service~1
-    return bufrw.str1.readFrom(frame.buffer, CallRequest.RW.lazy.serviceOffset);
+    var res = bufrw.str1.readFrom(
+        frame.buffer, CallRequest.RW.lazy.serviceOffset
+    );
+    frame.cache.serviceRes = res;
+    frame.cache.headerStartOffset = res.offset;
```
Comment (jcorbin) [inline v2/call.js]: maybe we can just cache serviceRes

File: v2/call.js [hunk 9]
```diff
@@ -96,37 +117,226 @@ CallRequest.RW.lazy.readTracing = function lazyReadTracing(frame) {
...
+        return null;
+    }
+    var serviceNameStr = fastBufferToString(
+        frame.buffer,
+        CallRequest.RW.lazy.serviceOffset + 1,
+        end
+    );
+
+    frame.cache.serviceStr = serviceNameStr;
+    frame.cache.headerStartOffset = end;
```
Comment (jcorbin) [inline v2/call.js]: please unify all the passages that get the headerStartOffset

File: v2/call.js [hunk 10]
```diff
@@ -96,37 +117,226 @@ CallRequest.RW.lazy.readTracing = function lazyReadTracing(frame) {
...
+            return null;
+        }
+        var strLength = frame.buffer.readUInt8(offset, false);
+        offset += strLength + 1;
+    }
+
+    if (frame.size < offset + 1) {
+        return null;
+    }
+    var nh = frame.buffer.readUInt8(offset, false);
```
Comment (jcorbin) [inline v2/call.js]: this header skipper / scanner should be (already is?) a separate thing

File: v2/call.js [hunk 11]
```diff
@@ -96,37 +117,226 @@ CallRequest.RW.lazy.readTracing = function lazyReadTracing(frame) {
...
+
+    return arg1Str;
+};
+
+function getHeadersOffset(frame) {
+    var res = null;
+    var offset = 0;
+
+    if (frame.cache.csumStartOffset) {
+        offset = frame.cache.csumStartOffset;
```
Comment (jcorbin) [inline v2/call.js]: could just be spelled `return ` rather than `offset =` and use short-circuiting

**Comment by jcorbin** [comment]
good stuff, time to break it up and refine imo

**Comment by Raynos** [comment]
This is now done & in.

**Pull request closed without merging**